### DISCUSSION TITLE
How to build mitm* binary from my own local source code

### DISCUSSION THREAD

**Discussion by BabyTigerChaseDream** [mitmproxy/mitmproxy#6068] [Q&A] [answered]
Below is steps I am following , but seems like none of my "print()" mitmproxy/tools/main.py got print out when I boot my 'self-built' binary located at **release/dist/**.

Binary building steps:
```
git clone https://github.com/mitmproxy/mitmproxy.git
cd mitmproxy
python3 -m venv venv
venv/bin/pip3 install -e ".[dev]"
source venv/bin/activate
python3 release/build.py standalone-binaries
```

print message I wrote in my code : tools/main.py

```
from __future__ import annotations

import argparse
import asyncio
import logging
import os
import signal
import sys
..... 
from mitmproxy.utils import debug

import traceback,inspect,time,logging
logger = logging.getLogger(__name__)

logger.log(logging.INFO, " ===== Hello BabyTiger ====== ")
```
Anything I missed ?
@mhils @cortesi

**Comment by mhils** [comment]
This looks about right!

**Reply by BabyTigerChaseDream** [reply]
But seems like when I boot my own mitmproxy binary locally , there is not logging in tools/main.py print out , I assume to see them on the terminal right ? tools/main.py is the entry point for the binary , so it is a 'must be called' code session

**Reply by BabyTigerChaseDream** [reply]
Or could you suggest how do you normally the code ? 
Use logging / print or is it possible to use **pdb** to set breakpoint ?
If possible could you share some detailed example on it ? @mhils

**Reply by mhils** [reply]
I typically don't build binaries locally, I just run mitmproxy from a regular Python environment.

Have you tried plain `print` instead of logging? You may be logging here before the logging handler is set up.

**Reply by BabyTigerChaseDream** [reply, marked as answer]
Hello Maximilian,
Things working right after I take following actions:
1. I quit (venv) 
2. re-entering venv by : "source ./venv/bin/activate"
3. re install all build dependency : ./venv/bin/pip install ".[dev]"
4. build binary again: "python release/build.py standalone-binaries"  , with logging posted in my question above 
5. when running selftest.py , I can see all my logging 
6. unzip binary : "cd release/dist/;tar -xf mitmproxy-dev-macos.tar.gz"
7. boot mitmdump : "./mitmdump -vvvv"

I got my logging on terminal now 
```
 ./mitmdump -vvvv 
 ===== Hello BabyTiger ====== 
```
